File: Accounting/src/main/java/com/experiment/accounting/service/base/EntityService.java
package com.experiment.accounting.service.base;

import com.experiment.accounting.entity.base.BaseEntity;
import com.experiment.accounting.repository.base.EntityRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Example;
import org.springframework.data.domain.ExampleMatcher;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class EntityService extends BaseService {

    protected final EntityRepository entityRepository;

    public EntityService(EntityRepository entityRepository){
        this.entityRepository = entityRepository;
    }

    public <T extends BaseEntity> T save(BaseEntity baseEntity) {
        if(baseEntity != null)
            return (T)entityRepository.save(baseEntity);
        return null;
    }

    public <T extends BaseEntity> Optional<T> find(Long id, Class<T> c) {
        ExampleMatcher matcher = ExampleMatcher.matchingAll();
        T instance = createInstance(c);
        instance.setId(id);
        if(id != null)
            return entityRepository.findOne(Example.of(instance, matcher));
        return Optional.empty();
    }

    public <T extends BaseEntity> List<T> findAll(Class<T> c) {
        T instance = createInstance(c);
        return entityRepository.findAll(Example.of(instance));
    }

    public <T extends BaseEntity> void delete(Long id, Class<T> c){
        T instance = createInstance(c);
        instance.setId(id);
        entityRepository.delete(instance);
    }
}

File: Inventory/src/main/kotlin/com/experiment/inventory/service/EntityService.kt
package com.experiment.inventory.service

import com.experiment.inventory.entity.base.BaseEntity
import com.experiment.inventory.repository.EntityRepository
import com.experiment.inventory.service.base.BaseService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.Example
import org.springframework.data.domain.ExampleMatcher
import org.springframework.stereotype.Service
import java.util.*

@Service
class EntityService<X:BaseEntity> @Autowired constructor(private var entityRepository: EntityRepository<X>): BaseService() {

    fun <T:X>save(baseEntity: T?): T? {
        if(baseEntity == null)
            return null
        return entityRepository.save(baseEntity)
    }

    fun <T:X> find(id: Long?, c: Class<T>): Optional<T> {
        if(id == null)
            return Optional.empty()
        val matcher = ExampleMatcher.matchingAll()
        val instance: T = createInstance(c)
        instance.id=id
        return entityRepository.findOne(Example.of(instance, matcher))
    }

    fun <T:X> findAll(c: Class<T>): List<T>? {
        val instance: T = createInstance(c)
        return entityRepository.findAll(Example.of(instance))
    }

    fun <T:X> delete(id: Long?, c: Class<T>) {
        val instance: T = createInstance(c)
        instance.id = id
        entityRepository.delete(instance)
    }
}